book/src/algorithm.md

```diff
@@ -0,0 +1,790 @@
+the way you resolve this package can change the requirements for your dependencies.
+
+## Checking Violations
+
+During AuditGraph construction violations are also checked. Violations have a [VersionReq][] and
+a list of violated criteria. They claim that, for all versions covered by the VersionReq, you believe
+that the listed criteria are explicitly violated. An error is produced if any edge is
+added to the AuditGraph where *either* endpoint matches the VersionReq and *any* criteria
+it claims to be an audit for is listed by the violation.
+
+This is an extremely complicated statement to parse, so let's look at some examples:
+
+```
+violation: safe-to-deploy, audit: safe-to-deploy -- ERROR!
+violation: safe-to-deploy, audit: safe-to-run    -- OK!
+violation: safe-to-run,    audit: safe-to-deploy -- ERROR!
+violation: [a, b],         audit: [a, c]         -- ERROR!
+```
+
+One very notable implication of this is that a violation for `["safe-to-run", "safe-to-deploy"]`
+is actually equivalent to `["safe-to-run"]`, not `["safe-to-deploy"]`! This means that the normal
+way of handling things, turning the violation's criteria into one CriteriaSet and checking
+if `audit.contains(violation)` is incorrect!
+
+We must instead do this check for each individual item in the violation:
+
+```rust
+let has_violation = violation.iter().any(|item| audit.contains(item));
+```
+
+It may seem a bit strange to produce an error if *any* audit is in any way contradicted
+by *any* violation. Is that necessary? Is that sufficient?
+
+It's definitely sufficient: it's impossible to validate a version without having an audit edge
+with an end-point in that version.
+
+I would argue that it's also *necessary*: the existence of any audit (or exemption)
+that is directly contradicted by a violation is essentially an integrity error on the
+claims that we are working with. Even if you don't even use the audit for anything
+anymore, people who are peering with you and importing your audits might be, so you
+should do something about those audits as soon as you find out they might be wrong!
+
+There is currently no mechanism for mechanically dealing with such an integrity error,
+even if the audit or violation comes from a foreign import. Such a situation is serious
+enough that it merits direct discussion between humans. That said, if this becomes
+enough of a problem we may eventually add such a feature.
+
+
+
+# Step 2a: Resolving Third Parties (Analyzing Audits)
+
+A lot of the heavy lifting for this task is in Step 1c (AuditGraph).
+
+Trying to validate all criteria at once is slightly brain-melty (because
+different criteria may be validated by different paths), so as a simplifying
+step we validate each criteria individually (so everything I'm about to
+describe happens in a for loop).
+
+If all we care about is finding out if a package has some criteria, then all
+we need to do is run depth-first-search ([DFS][]) from the Root Node and see if it reaches

```

Comment:
I'm not sure if the code is technically a depth-first search anymore given that it's now using a priority queue rather than a stack (since #296). Heaps don't preserve node order in the same way that a stack would.

I suppose it's more of a [Best First Search](https://en.wikipedia.org/wiki/Best-first_search) but with the edge weights simplified because no successor can have better hereustics than its parent, and the hereustics have nothing to do with distance to target (though that is a potential thing we could do, by adding a distance into the sorting key)